File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/text_charAt.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeTextCharAt(statement: JSONObject, onError: (e: Exception) -> Unit): String? {
    try {
        Log.d("hello", "execute_text_charAt")
        val pointer = statement.getJSONObject("pointer")
        val value = executeStatement(pointer["value"].toString())?.toString() ?: return null
        val at = executeStatement(pointer["at"].toString())?.toString()?.toInt() ?: return null

        return when (pointer["action"].toString()) {
            "FROM_START" -> value[at - 1].toString()
            "FROM_END" -> value[value.length - at].toString()
            "FIRST" -> value.first().toString()
            "LAST" -> value.last().toString()
            "RANDOM" -> value.random().toString()
            else -> null
        }
    } catch (e: Exception) {
        onError(e)
    }
    return null
}

File: piper/extensions/chrome/src/engine/statements/text_charAt.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let value = await this.executeStatement(stringify(pointer.value))
        value = stringify(value)
        let at = await this.executeStatement(stringify(pointer.at))
        
        switch(stringify(pointer.action)) {
            case "FROM_START":
                at = parseInt(at)
                return value[at - 1]
            case "FROM_END":
                at = parseInt(at)
                return value[value.length - at]
            case "FIRST":
                return value[0]
            case "LAST":
                return value[value.length - 1]
            case "RANDOM":
                return value[Math.floor(Math.random() * value.length)]
        }
    } catch (e) {
        console.log(e)
        onError(e)
    }

    return null
}
